stan/math/prim/prob/gaussian_dlm_obs_lpdf.hpp

```diff
@@ -61,58 +63,60 @@ namespace math {
-    const Eigen::Matrix<T_W, Eigen::Dynamic, Eigen::Dynamic>& W,
-    const Eigen::Matrix<T_m0, Eigen::Dynamic, 1>& m0,
-    const Eigen::Matrix<T_C0, Eigen::Dynamic, Eigen::Dynamic>& C0) {
-  using T_lp
-      = return_type_t<T_y, return_type_t<T_F, T_G, T_V, T_W, T_m0, T_C0>>;
+          typename T_W, typename T_m0, typename T_C0,
+          require_all_eigen_matrix_t<T_y, T_F, T_G, T_V, T_W, T_C0>* = nullptr,
+          require_eigen_col_vector_t<T_m0>* = nullptr>
+inline return_type_t<T_y, T_F, T_G, T_V, T_W, T_m0, T_C0> gaussian_dlm_obs_lpdf(
+    const T_y& y, const T_F& F, const T_G& G, const T_V& V, const T_W& W,
+    const T_m0& m0, const T_C0& C0) {
+  using T_lp = return_type_t<T_y, T_F, T_G, T_V, T_W, T_m0, T_C0>;
   using std::pow;
-  int r = y.rows();  // number of variables
-  int T = y.cols();  // number of observations
-  int n = G.rows();  // number of states
-
   static const char* function = "gaussian_dlm_obs_lpdf";
-  check_finite(function, "y", y);
-  check_not_nan(function, "y", y);
   check_size_match(function, "columns of F", F.cols(), "rows of y", y.rows());
   check_size_match(function, "rows of F", F.rows(), "rows of G", G.rows());
-  check_finite(function, "F", F);
-  check_square(function, "G", G);
-  check_finite(function, "G", G);
   check_size_match(function, "rows of V", V.rows(), "rows of y", y.rows());
-  // TODO(anyone): incorporate support for infinite V
-  check_finite(function, "V", V);
-  check_pos_semidefinite(function, "V", V);
   check_size_match(function, "rows of W", W.rows(), "rows of G", G.rows());
-  // TODO(anyone): incorporate support for infinite W
-  check_finite(function, "W", W);
-  check_pos_semidefinite(function, "W", W);
   check_size_match(function, "size of m0", m0.size(), "rows of G", G.rows());
-  check_finite(function, "m0", m0);
   check_size_match(function, "rows of C0", C0.rows(), "rows of G", G.rows());
-  check_pos_definite(function, "C0", C0);
-  check_finite(function, "C0", C0);
+  check_square(function, "G", G);
+
+  const auto& y_ref = to_ref(y);
+  const auto& F_ref = to_ref(F);
+  const auto& G_ref = to_ref(G);
+  const auto& V_ref = to_ref(V);
+  const auto& W_ref = to_ref(W);
+  const auto& m0_ref = to_ref(m0);
+  const auto& C0_ref = to_ref(C0);
+
+  check_finite(function, "y", y_ref);
+  check_finite(function, "F", F_ref);
+  check_finite(function, "G", G_ref);
+  // TODO(anyone): incorporate support for infinite V
+  check_finite(function, "V", V_ref);
+  check_pos_semidefinite(function, "V", V_ref);
+  // TODO(anyone): incorporate support for infinite W
+  check_finite(function, "W", W_ref);
+  check_pos_semidefinite(function, "W", W_ref);
+  check_finite(function, "m0", m0_ref);
+  check_pos_definite(function, "C0", C0_ref);
+  check_finite(function, "C0", C0_ref);
 
   if (size_zero(y)) {
     return 0;
   }
 
+  int r = y.rows();  // number of variables
+  int T = y.cols();  // number of observations
+  int n = G.rows();  // number of states
+
```

Comment:
If I am reading this right, T and r are only used in once place. Do they have to be a local variable?

 But you just moved this code around so this is optional.

Reply:
r is used in a number of places, so I will leave it be. I can eliminate T tho.